design/construct-tree.md

```diff
@@ -0,0 +1,343 @@
+
+### Construct Tree Data Model
+
+A feature will be added to the CDK core module so that any CDK application will produce an artifact called `tree.json` during calls to `cdk synth` and include it as an output to the cloud assembly in `cdk.out`. Developers will be able to opt-out if this functionality is not desired or needed through a flag in the CLI or set it as a preference.
+
+The construct tree will be a list of paths that are indexed into a flat map of constructs. The file structure itself represents the tree. This information will need to include the following to start rendering a flat map of constructs:
+
+### Construct properties
+
+|Property    |Type    |Required |Description  |
+|---  |---  |---  |---  |
+|id |string |Required |id of the construct within the current scope |
+|path |string |Required |full, absolute path of the construct within the tree |
+|children |Array<Path>  |Not Required |All direct children of this construct. Array of the absolute paths of the constructs. Will be used to walk entire list of constructs |
+|metadata |Array<Metadata>  |Not Required |Metadata describing all constructs/resources that are encapsulated by the construct |
+
+### Metadata Properties
+
+The following metadata properties will be included by the construct that produces the `tree.json` output.
+
+|Property    |Type    |Required    |Description    |
+|---    |---    |---    |---  |
+|sourceLocation    |string    |Required    |location in source code where the construct is defined    |
+|resourceType |string |Required |type of resource (schema / classification terminology TBD) |
+|description    |string |Not Required |description of the construct |
+|renderedView    |Array<Metadata> |Not Required    |constructs can fill in arbitrary. ClouFormation agnostic. CFN constructs could pre-populate it with the output of `this.toCloudFormation()` |
+|errors |Array<Metadata>  |Not Required |array of errors associated with this construct.  |
+|warnings |Array<Metadata>  |Not Required |array of warnings associated with this construct |
+
+This is a temporary placeholder (WIP) of the construct tree in the JSON schema format described above. It’s a representation of the tree for the prototype referenced earlier in the Overview section of the CDK app in the [cdkworkshop](https://cdkworkshop.com/)
+
+**TODO** - add detail and a more concrete walkthrough with samples of all the attributes referenced in construct and metadata properties.
+
+```json
+{
+  "id": "App",
+  "path": "",
+  "metadata": {
+    "sourceLocation": "",
+    "resourceType": "AWS::CDK::App"
+  },
+  "children": [
+    {
+      "id": "CdkWorkshopStack",
+      "path": "CdkWorkshopStack",
+      "resourceType": "AWS::CloudFormation::Stack",
```

Comment:
did you mean the stack itself or the `AWS::CloudFormation::Stack` type? The reason I included the stack is because it has properties and configured values that might be useful to quickly look up from the construct tree.

It's the resource that nests a CloudFormation [stack](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-stack.html) in a top level template.

Reply:
The `core.Stack` class does not represent really the `AWS::CloudFormation::Stack` CloudFormation resource, which is a nested stack.

Reply:
+1. See my comment above in the requirements section around types & hierarchy. I'm not convinced we know what we want here.